src/upgrade.py

```diff
@@ -0,0 +1,115 @@
+    DataUpgrade,
+    DependencyModel,
+    UpgradeGrantedEvent,
+    verify_requirements,
+)
+from pydantic import BaseModel
+from typing_extensions import override
+
+from utils import get_zookeeper_version
+
+if TYPE_CHECKING:
+    from charm import KafkaCharm
+
+logger = logging.getLogger(__name__)
+
+
+class KafkaDependencyModel(BaseModel):
+    """Model for Kafka Operator dependencies."""
+
+    service: DependencyModel
+
+
+class KafkaUpgrade(DataUpgrade):
+    """Implementation of :class:`DataUpgrade` overrides for in-place upgrades."""
+
+    def __init__(self, charm: "KafkaCharm", **kwargs):
+        super().__init__(charm, **kwargs)
+        self.charm = charm
+
+    @property
+    def idle(self) -> bool:
+        """Checks if cluster state is idle.
+
+        Returns:
+            True if cluster state is idle. Otherwise False
+        """
+        return self.cluster_state == "idle"
+
+    @property
+    def current_version(self) -> str:
+        """Get current Kafka version."""
+        dependency_model: DependencyModel = getattr(self.dependency_model, "service")
+        return dependency_model.version
+
+    @property
+    def zookeeper_current_version(self) -> str:
+        """Get current Zookeeper version."""
+        return get_zookeeper_version(zookeeper_config=self.charm.kafka_config.zookeeper_config)
+
+    @override
+    def pre_upgrade_check(self) -> None:
+        default_message = "Pre-upgrade check failed and cannot safely upgrade"
+        if not self.charm.healthy:
+            raise ClusterNotReadyError(message=default_message, cause="Cluster is not healthy")
+
+    @override
+    def build_upgrade_stack(self) -> list[int]:
+        upgrade_stack = []
+        units = set([self.charm.unit] + list(self.charm.peer_relation.units))  # type: ignore[reportOptionalMemberAccess]
+        for unit in units:
+            upgrade_stack.append(int(unit.name.split("/")[-1]))
+
+        return upgrade_stack
+
+    @override
+    def log_rollback_instructions(self) -> None:
+        logger.warning("SOME USEFUL INSTRUCTIONS")  # TODO
+
+    @override
+    def _on_upgrade_granted(self, event: UpgradeGrantedEvent) -> None:
+        dependency_model: DependencyModel = getattr(self.dependency_model, "service")
+        if not verify_requirements(
+            version=self.zookeeper_current_version,
+            requirement=dependency_model.dependencies["zookeeper"],
+        ):
+            logger.error("ZooKeeper requirement not met")
+            self.set_unit_failed()
+            return
+
+        self.charm.snap.stop_snap_service()
+
+        if not self.charm.snap.install():
+            logger.error("Unable to install Snap")
+            self.set_unit_failed()
+            return
+
+        logger.info(f"{self.charm.unit.name} upgrading service...")
+        self.charm.snap.restart_snap_service()
+
+        try:
+            logger.debug("Running post-upgrade check...")
+            self.pre_upgrade_check()
```

Comment:
**todo** I believe it is clearer to refer to then as post_upgrade check. I think I sort of raised a similar point in Kafka, [here](https://github.com/canonical/zookeeper-operator/pull/71/files#diff-b0d4c16f2b702229d288067c4933254e190e3c0ac6696e337e604b33f970f47dR139). It would be good to keep the same naming here too

Reply:
Yes. This could be added to the `upgrade` lib probably, so it follows same pattern for being `@abstractmethod` on `DataUpgrade` class maybe? @marcoppenheimer thoughts?

Reply:
I think it makes sense for Kafka+ZK as we have methods already in place for doing these 'health' checks. So our pre + post look the same. 
That might not be true for other products, so I would be against adding it in to the base `DataUpgrade`.